bigframes/operations/semantics.py

```diff
@@ -527,6 +527,109 @@ def search(
+            animal  animal_1
+            0  monkey    baboon
+            1  spider  scorpion
+            <BLANKLINE>
+            [2 rows x 2 columns]
+
+        Args:
+            other (DataFrame):
+                The other data frame to join with.
+            left_on (str):
+                The name of the column on left side for the join.
+            right_on (str):
+                The name of the column on the right side for the join.
+            top_k (int, default 3):
+                The number of nearest neighbors to return.
+            model (TextEmbeddingGenerator):
+                A TextEmbeddingGenerator provided by Bigframes ML package.
+            score_column (Optional[str], default None):
+                The name of the the additional column containning the similarity scores. If None,
+                this column won't be attached to the result.
+            max_rows:
+                The maximum number of rows allowed to be processed per call. If the result is too large, the method
+                call will end early with an error.
+
+        Returns:
+            DataFrame: the data frame with the join result.
+
+        Raises:
+            ValueError: when the amount of data to be processed exceeds the specified max_rows.
+        """
+
+        if left_on not in self._df.columns:
+            raise ValueError(f"Left column {left_on} not found")
+        if right_on not in self._df.columns:
+            raise ValueError(f"Right column {right_on} not found")
+
+        import bigframes.ml.llm as llm
+
+        if not isinstance(model, llm.TextEmbeddingGenerator):
+            raise TypeError(f"Expect a text embedding model, but got: {type(model)}")
+
+        joined_table_rows = len(self._df) * len(other)
+        if joined_table_rows > max_rows:
+            raise ValueError(
+                f"Number of rows that need processing is {joined_table_rows}, which exceeds row limit {max_rows}."
+            )
+
+        base_table = self._attach_embedding(other, right_on, model).to_gbq()
+        query_table = self._attach_embedding(self._df, left_on, model)
+
+        import bigframes.bigquery as bbq
+
+        join_result = bbq.vector_search(
+            base_table=base_table,
+            column_to_search="embedding",
+            query=query_table,
+            top_k=top_k,
+        )
+
+        join_result = join_result.drop(["embedding", "embedding_1"], axis=1)
+
+        if score_column is not None:
+            join_result = join_result.rename(columns={"distance": score_column})
+        else:
+            del join_result["distance"]
+
+        return join_result
+
+    @staticmethod
+    def _attach_embedding(dataframe, column: str, model):
+        result_df = dataframe.copy()
+        embeddings = model.predict(dataframe[column])["ml_generate_embedding_result"]
+        result_df["embedding"] = embeddings
```

Comment:
To prevent naming conflicts in the agg method, if the dataframe already contains a column named embedding, the aggregation_group_id column should be assigned a unique identifier. This can be achieved by using a UUID generator like `bigframes.core.guid.generate_guid("agg")`.

Reply:
Good call. Though for the query table I have to hard code the "embedding" column.